### PR TITLE
Fix math for undiscovered species boosts

### PR DESCRIPTION
**Pull request by jeremyd85**
* Add boosted_random_choice to be used for reseach boosts
* Add tests for probability code
* Add _recruit_in_tier function to select the specific species

### CODE REVIEW

File: tests/utils/test_utils.py [hunk 1]
```diff
@@ -13,3 +17,26 @@ def test_fibonacci_sum():
...
+def test_boosted_random_choice(arr, boost_percent, expected_percent):
+    random.seed(42)
+    boosted_count = 0
+    total = 10000
+    for _ in range(total):
+        choice = boosted_random_choice(arr, lambda x: x < 3, boost_percent)
+        if choice < 3:
+            boosted_count += 1
+    percent_boosted = boosted_count / total
+    assert expected_percent - 0.01 < percent_boosted < expected_percent + 0.01
```
Comment (rj-lee) [inline tests/utils/test_utils.py:42]:
```suggestion
    assert abs(percent_boosted - expected_percent) < 0.01
```
Assert that the absolute difference is less than this threshold for readability++

**Comment by TheRedPanda17** [comment]
How much have you tested this locally?

File: myning/utils/species_rarity.py [hunk 2]
```diff
@@ -38,7 +40,23 @@
 SPECIES_WEIGHTS = [150, 75, 40, 20, 10, 7, 4]
 
 
-def get_recruit_species(highest_rarity: int):
+def _recruit_in_tier(tier: list[CharacterSpecies]) -> Optional[Species]:
+    player = Player()
+
+    def is_undiscovered(species_name: CharacterSpecies):
+        return SPECIES[species_name] not in player.discovered_species
```
Comment (rj-lee) [inline myning/utils/species_rarity.py:47]: could be defined outside this function, although you would need to add another `player = Player()`

File: myning/utils/utils.py [hunk 3]
```diff
@@ -38,6 +38,37 @@ def get_random_array_item(arr: ListType[T]) -> T:
     return get_random_array_item_and_index(arr)[0]
 
 
+def boosted_random_choice(
+    arr: list[T], selector: Callable[[T], bool], percent_boost: float = 0.0
+) -> Optional[T]:
```
Comment (rj-lee) [inline myning/utils/utils.py:43]:
```suggestion
) -> T | None:
```

I have a preference for the union type operator that I'd like to share with you 😆 

`Optional[T]` sounds more like the thing to use when defining a dataclass with optional properties

**Review by rj-lee — COMMENTED** [review]
Was mostly looking at code quality

File: myning/utils/utils.py (hunk 3 above)
Comment (jeremyd85) [reply myning/utils/utils.py:43]: I never thought of this! Thanks RJ for the enlightenment!

File: myning/utils/species_rarity.py (hunk 2 above)
Comment (jeremyd85) [reply myning/utils/species_rarity.py:47]: What do you think of just a named lambda function?

**Comment by jeremyd85** [comment]
> How much have you tested this locally?

I haven't. I'll switch to it until I get a companion just to make sure there is no exception thrown.

**Comment by jeremyd85** [comment]
Human recruited!

File: myning/utils/species_rarity.py (hunk 2 above)
Comment (rj-lee) [reply myning/utils/species_rarity.py]: well the point is that it's not necessary to redefine this function on each call of the outer function, so might as well just make a regular function

**Comment by TheRedPanda17** [comment]
I'm cool with the code, but I want it tested on a save that has undiscovered species available in a mine. @rj-lee  do you have any save files that fit this description?

**Comment by rj-lee** [comment]
> I'm cool with the code, but I want it tested on a save that has undiscovered species available in a mine. @rj-lee do you have any save files that fit this description?

I do, but everyone also should lol nobody has unlocked everything yet

**Comment by TheRedPanda17** [comment]
> > I'm cool with the code, but I want it tested on a save that has undiscovered species available in a mine. @rj-lee do you have any save files that fit this description?
> 
> I do, but everyone also should lol nobody has unlocked everything yet

I don't. I went back in time, and before that, I had unlocked all the races for the mines I had available.

**Comment by jeremyd85** [comment]
I vote we ship it!

**Review by TheRedPanda17 — APPROVED** [review]

**Pull request merged**